tryouts/transaction.ts

```diff
@@ -0,0 +1,102 @@
+export interface IObject<T = any> {
+    [propery: string]: T;
+}
+
+class Store {
+  public openSession(): Session;
+  public async openSession(transaction: (session: Session) => Promise<void>): Promise<void>;
+  public openSession(transaction?: (session: Session) => Promise<void>): Session | Promise<void> {
+    const session: Session = new Session();
+    
+    if (!transaction) {
+      return session;
+    }
+
+    return transaction(session.initialize())
+      .then(() => session.saveChanges())
+      .then(() => this.finalize());
+  }
+
+  public async finalize(): Promise<void> {
+    console.log('finalize');
+    return Promise.resolve();
+  }
+}
+
+class Session {
+  protected initialized = false;
+
+  public initialize(): Session {
+    console.log('initialize');
+    this.initialized = true;
+
+    return this;
+  }
+
+  public create<T extends object = IObject>(json: object): T {
+    this.assertInitialize();
+    return json as T; 
+  }
+
+  public async load<T extends object = IObject>(id: string): Promise<T> {    
+    let entity: T = {} as T;
+    (entity as IObject).id = parseInt(id);
+
+    this.assertInitialize();    
+    console.log('load', id);
+    return Promise.resolve<T>(entity);
+  }
+
+  public async store<T extends object = IObject>(entity: T): Promise<T> {
+    this.assertInitialize();
+    console.log('store', entity);
+    return Promise.resolve<T>(entity);
+  }
+
+  public async saveChanges(): Promise<void> {
+    this.assertInitialize();
+    console.log('save changes');
+    return Promise.resolve();
+  }
+
+  protected assertInitialize(): void {
+    if (!this.initialized) {
+      throw new Error('Not initialized');
+    }
+  }
+}
+
+class Runner
+{
+  public static async run(): Promise<void> {
+    const store = new Store();
+    const session = store.openSession();
+    let doc1: IObject, doc2: IObject;
+
+    session.initialize();
+    doc1 = session.create({id:1});
+    doc2 = await session.load('2');
+
+    doc1 = await session.store(doc1);
+    doc2 = await session.store(doc2);    
+
+    await session.saveChanges();
+    await store.finalize();
+
+    console.log('doc1', doc1);
+    console.log('doc2', doc2);
+
+    await store.openSession(async (session: Session): Promise<void> => {
```

Comment:
> people which are accustomed to this workflow will be confused why changes won't apply.

I think they'll learn soon enough ;)

Coming from C# client I thought it's made this way to make up for the lack of `using()` clause in JS, but since there's nothing to dispose of, then we probably can avoid that. Also I think we don't want to be like others, but be consistent across different ecosystems.

On the other hand I can see the convenience of this. And there's another role of `using()` (grouping/closing a bunch of calls in some context). Since we don't have a `using()` to group our session calls for clearness sake it might have sense to introduce this callback accepting method in this client.

To sum up. I'd say we keep the callback version, **but** don't do `saveChanges()` automatically. This way it's just grouping the session calls, but that's OK. It's clear what's done with the session and when it's usage ends. Like this:

```javascript
await store.openSession(async (session: Session): Promise<void> => {
      doc1 = session.create({id:1});
      doc2 = await session.load('2');

      doc1 = await session.store(doc1);
      doc2 = await session.store(doc2);
      await session.saveChanges();
    });
```

Thoughts guys?